Create a filtermodel decorator

zachmullen opened the pull request:
@mgrauer PTALSTWYIO. If you are OK with how this looks, I will go ahead and convert all of our REST endpoints over. This will be a good exercise; in the folder resource alone I found two endpoints that were failing to filter their results.

mgrauer commented:
> PTALSTWYIO

See if This is What You Intended Originally?

Such That Walrus Yawping Is Obstreperous?


Yes to both.  I like it.

zachmullen commented:
> it seems like I am adding fields to a filter.

You are adding fields to the filter. The filter says what fields are allowed in the document, and you're adding other fields to that filter... maybe the word "filter" itself is a source of confusion here?

mgrauer commented:
Indeed, I think of "filter" as what is removed. _.js uses "pluck" for the operation of passing through certain selected values.

I can live with it if the docstring clarified a bit more.

zachmullen commented:
Ah I see. I'll add more to the docstring to indicate that this is the "whitelist" sort of filter. I'm used to thinking of a filter as specifying what passes through rather than what doesn't, but I can definitely see the ambiguity.

zachmullen commented:
> PTALSTWYIO

Since this was your idea originally. I think I've addressed all your comments, so if you like what you see I will convert all the existing endpoints over.

mgrauer commented:
LGTM.  Go for it.

:checkered_flag:

zachmullen commented:
@girder/developers PTAL at this change, this will touch a lot of REST code. It's another step towards reducing a lot of our boilerplate into a set of decorator annotations.

jcfr commented:
:+1:

zachmullen commented:
Everything I care to convert for the moment has been converted.

zachmullen commented:
Fixed.

mgrauer commented on girder/models/item.py:
```diff
@@ -339,6 +347,11 @@ def parentsToRoot(self, item, user=None, force=False):
 
         :param item: The item whose root to find
         :type item: dict
+        :param user: The user running the test (not required if force=True).
```
`running the test` ?  seems out of place here

zachmullen replied:
This is one of those "did I really write that?" moments.

mgrauer commented:
I'm wondering if there is an inconsistency in when you add the `access` to the filter, but looking through the code, I couldn't find any.  There's a lot of code though.  Can you summarize the policy for when to add `access` to the filter?

zachmullen commented:
We more or less throw it into the document for any request where it could have changed. This minimizes the number of requests that client needs to make to the server. (At least, that's the goal... the execution might not be 100% correct.)

mgrauer commented:
> Everything I care to convert for the moment has been converted.

    f(sizeof(codebase), sizeof(your_heart))

LGTM.
